File: benchmarks/lib/system.ts
import { readFile } from "node:fs/promises";
import { arch, cpus, hostname, platform, release, totalmem } from "node:os";

export async function sha256(bytes: Uint8Array): Promise<string> {
  const digest = new Uint8Array(
    await crypto.subtle.digest("SHA-256", bytes as Uint8Array<ArrayBuffer>),
  );
  return [...digest].map((byte) => byte.toString(16).padStart(2, "0")).join("");
}

async function maybeFile(path: string): Promise<string | undefined> {
  try {
    return await readFile(path, "utf8");
  } catch {
    return undefined;
  }
}

function procKiB(text: string | undefined, key: string): number | undefined {
  const match = text?.match(new RegExp(`^${key}:\\s+(\\d+)\\s+kB$`, "m"));
  return match ? Number(match[1]) * 1024 : undefined;
}

export async function processMemory(): Promise<Record<string, number>> {
  const [status, rollup] = await Promise.all([
    maybeFile("/proc/self/status"),
    maybeFile("/proc/self/smaps_rollup"),
  ]);
  const values = {
    rssBytes: procKiB(status, "VmRSS"),
    peakRssBytes: procKiB(status, "VmHWM"),
    pssBytes: procKiB(rollup, "Pss"),
  };
  return Object.fromEntries(
    Object.entries(values).filter((entry): entry is [string, number] => entry[1] !== undefined),
  );
}

/** Host pressure / frequency metadata for PERF-013. Best-effort; missing fields are omitted. */
export async function hostPerfMetadata(): Promise<Record<string, unknown>> {
  const [governor, scalingCur, scalingMax, vmstat, pressure] = await Promise.all([
    maybeFile("/sys/devices/system/cpu/cpu0/cpufreq/scaling_governor"),
    maybeFile("/sys/devices/system/cpu/cpu0/cpufreq/scaling_cur_freq"),
    maybeFile("/sys/devices/system/cpu/cpu0/cpufreq/scaling_max_freq"),
    maybeFile("/proc/vmstat"),
    maybeFile("/proc/pressure/cpu"),
  ]);
  const out: Record<string, unknown> = {};
  if (governor) out.cpuGovernor = governor.trim();
  if (scalingCur) out.cpuFreqKhz = Number(scalingCur.trim());
  if (scalingMax) out.cpuMaxFreqKhz = Number(scalingMax.trim());
  if (vmstat) {
    const pgmaj = vmstat.match(/^pgmajfault\s+(\d+)$/m);
    const pswpin = vmstat.match(/^pswpin\s+(\d+)$/m);
    const pswpout = vmstat.match(/^pswpout\s+(\d+)$/m);
    if (pgmaj) out.majorFaults = Number(pgmaj[1]);
    if (pswpin) out.swapIn = Number(pswpin[1]);
    if (pswpout) out.swapOut = Number(pswpout[1]);
  }
  if (pressure) {
    const some = pressure.match(/^some avg10=([0-9.]+)/m);
    if (some) out.cpuPressureAvg10 = Number(some[1]);
  }
  return out;
}

export async function systemMetadata(): Promise<Record<string, unknown>> {
  const cpuList = cpus();
  return {
    hostname: hostname(),
    os: platform(),
    osRelease: release(),
    architecture: arch(),
    logicalCpus: cpuList.length,
    cpuModel: cpuList[0]?.model ?? "unknown",
    totalMemoryBytes: totalmem(),
    runtime: `${process.release.name} ${process.version}`,
    process: await processMemory(),
    perf: await hostPerfMetadata(),
  };
}

export function monotonicMs(): number {
  return performance.now();
}

File: benchmarks/runners/native/system.rs
use std::fs;
use std::process::Command;

pub(crate) fn proc_bytes(key: &str) -> Option<u64> {
    let status = fs::read_to_string("/proc/self/status").ok()?;
    let line = status.lines().find(|line| line.starts_with(key))?;
    let kib = line.split_whitespace().nth(1)?.parse::<u64>().ok()?;
    Some(kib * 1024)
}

pub(crate) fn pss_bytes() -> Option<u64> {
    let rollup = fs::read_to_string("/proc/self/smaps_rollup").ok()?;
    let line = rollup.lines().find(|line| line.starts_with("Pss:"))?;
    Some(line.split_whitespace().nth(1)?.parse::<u64>().ok()? * 1024)
}

pub(crate) fn command_output(program: &str, args: &[&str]) -> Option<String> {
    let output = Command::new(program).args(args).output().ok()?;
    output
        .status
        .success()
        .then(|| String::from_utf8_lossy(&output.stdout).trim().to_owned())
}

pub(crate) fn host_perf_metadata() -> serde_json::Value {
    let governor = fs::read_to_string("/sys/devices/system/cpu/cpu0/cpufreq/scaling_governor")
        .ok()
        .map(|s| s.trim().to_owned());
    let freq = fs::read_to_string("/sys/devices/system/cpu/cpu0/cpufreq/scaling_cur_freq")
        .ok()
        .and_then(|s| s.trim().parse::<u64>().ok());
    let max_freq = fs::read_to_string("/sys/devices/system/cpu/cpu0/cpufreq/scaling_max_freq")
        .ok()
        .and_then(|s| s.trim().parse::<u64>().ok());
    let vmstat = fs::read_to_string("/proc/vmstat").unwrap_or_default();
    let pick = |key: &str| {
        vmstat.lines().find_map(|line| {
            let mut parts = line.split_whitespace();
            (parts.next() == Some(key))
                .then(|| parts.next()?.parse::<u64>().ok())
                .flatten()
        })
    };
    serde_json::json!({
        "cpuGovernor": governor,
        "cpuFreqKhz": freq,
        "cpuMaxFreqKhz": max_freq,
        "majorFaults": pick("pgmajfault"),
        "swapIn": pick("pswpin"),
        "swapOut": pick("pswpout"),
        "perfTracing": perf_enabled(),
    })
}

pub(crate) fn system_metadata() -> serde_json::Value {
    let cpuinfo = fs::read_to_string("/proc/cpuinfo").unwrap_or_default();
    let cpu_model = cpuinfo
        .lines()
        .find_map(|line| line.strip_prefix("model name\t: "))
        .unwrap_or("unknown");
    let mem_total = fs::read_to_string("/proc/meminfo")
        .ok()
        .and_then(|text| {
            text.lines()
                .find(|line| line.starts_with("MemTotal:"))
                .and_then(|line| line.split_whitespace().nth(1))
                .and_then(|value| value.parse::<u64>().ok())
        })
        .map(|kib| kib * 1024);
    serde_json::json!({
        "hostname": command_output("hostname", &[]).unwrap_or_else(|| "unknown".to_owned()),
        "os": std::env::consts::OS,
        "architecture": std::env::consts::ARCH,
        "logicalCpus": std::thread::available_parallelism().map(|n| n.get()).unwrap_or(1),
        "cpuModel": cpu_model,
        "totalMemoryBytes": mem_total,
        "runtime": "wasmtime via memcontainers/hosts/wasmtime",
        "process": {
            "rssBytes": proc_bytes("VmRSS:"),
            "peakRssBytes": proc_bytes("VmHWM:"),
            "pssBytes": pss_bytes()
        },
        "perf": host_perf_metadata(),
    })
}

/// Opt-in PERF-013 tracing. Accepts `1` / `true` / `yes` (case-insensitive). Unset or other values = off.
pub(crate) fn perf_enabled() -> bool {
    match std::env::var("MC_PERF") {
        Ok(v) => {
            let v = v.trim();
            v == "1" || v.eq_ignore_ascii_case("true") || v.eq_ignore_ascii_case("yes")
        }
        Err(_) => false,
    }
}
